### PR TITLE
feat(planner): plan table projection

### PR DESCRIPTION
**Pull request by skyzh**
Signed-off-by: Alex Chi <iskyzh@gmail.com>

This PR refactors the planner, and adds support for planning table projection. e.g.,

```
select colA from __mock_table_1;
```

Previously we only support `select *`.

* Support bind varchar.
* Add skeleton for filter, projection, sort, topN, values executor.
* Remove distinct executor / close https://github.com/cmu-db/bustub/issues/258

### CODE REVIEW

File: src/execution/filter_executor.cpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#include "execution/executors/filter_executor.h"
+
+namespace bustub {
+
+FilterExecutor::FilterExecutor(ExecutorContext *exec_ctx, const FilterPlanNode *plan,
+                               std::unique_ptr<AbstractExecutor> &&child_executor)
+    : AbstractExecutor(exec_ctx) {}
+
+void FilterExecutor::Init() {}
```
Comment (lmwnshn) [inline src/execution/filter_executor.cpp:9]: Thoughts on making skeletons raise exceptions saying they're not implemented in Init?

File: src/include/execution/executors/filter_executor.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+namespace bustub {
+
+/**
+ * The FilterExecutor executor executes a filter.
+ */
+class FilterExecutor : public AbstractExecutor {
+ public:
+  /**
+   * Construct a new SeqScanExecutor instance.
```
Comment (lmwnshn) [inline src/include/execution/executors/filter_executor.h:31]:
```suggestion
   * Construct a new FilterExecutor instance.
```

File: src/include/execution/executors/filter_executor.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+   * @param exec_ctx The executor context
+   * @param plan The sequential scan plan to be executed
```
Comment (lmwnshn) [inline src/include/execution/executors/filter_executor.h:33]:
```suggestion
   * @param plan The filter plan node to be executed
```

File: src/include/execution/executors/projection_executor.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace bustub {
+
+/**
+ * The ProjectionExecutor executor executes a projection.
+ */
+class ProjectionExecutor : public AbstractExecutor {
+ public:
+  /**
+   * Construct a new SeqScanExecutor instance.
```
Comment (lmwnshn) [inline src/include/execution/executors/projection_executor.h:31]:
```suggestion
   * Construct a new ProjectionExecutor instance.
```

File: src/include/execution/executors/projection_executor.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+   * @param exec_ctx The executor context
+   * @param plan The sequential scan plan to be executed
```
Comment (lmwnshn) [inline src/include/execution/executors/projection_executor.h:33]:
```suggestion
   * @param plan The projection plan to be executed
```

File: src/include/execution/executors/sort_executor.h [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace bustub {
+
+/**
+ * The SortExecutor executor executes a sort.
+ */
+class SortExecutor : public AbstractExecutor {
+ public:
+  /**
+   * Construct a new SeqScanExecutor instance.
```
Comment (lmwnshn) [inline src/include/execution/executors/sort_executor.h:31]:
```suggestion
   * Construct a new SortExecutor instance.
```

File: src/include/execution/executors/sort_executor.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+   * @param exec_ctx The executor context
+   * @param plan The sequential scan plan to be executed
```
Comment (lmwnshn) [inline src/include/execution/executors/sort_executor.h:33]:
```suggestion
   * @param plan The sort plan to be executed
```

File: src/include/execution/executors/topn_executor.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace bustub {
+
+/**
+ * The TopNExecutor executor executes a topn.
+ */
+class TopNExecutor : public AbstractExecutor {
+ public:
+  /**
+   * Construct a new SeqScanExecutor instance.
```
Comment (lmwnshn) [inline src/include/execution/executors/topn_executor.h:31]:
```suggestion
   * Construct a new TopNExecutor instance.
```

File: src/include/execution/executors/topn_executor.h [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+   * @param exec_ctx The executor context
+   * @param plan The sequential scan plan to be executed
```
Comment (lmwnshn) [inline src/include/execution/executors/topn_executor.h:33]:
```suggestion
   * @param plan The topn plan to be executed
```

File: src/include/execution/executors/values_executor.h [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace bustub {
+
+/**
+ * The ValuesExecutor executor produces rows of values.
+ */
+class ValuesExecutor : public AbstractExecutor {
+ public:
+  /**
+   * Construct a new SeqScanExecutor instance.
```
Comment (lmwnshn) [inline src/include/execution/executors/values_executor.h:31]:
```suggestion
   * Construct a new ValuesExecutor instance.
```

File: src/include/execution/executors/filter_executor.h [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+  /**
+   * Yield the next tuple from the filter.
+   * @param[out] tuple The next tuple produced by the filter
+   * @param[out] rid The next tuple RID produced by the filter
+   * @return `true` if a tuple was produced, `false` if there are no more tuples
+   */
+  auto Next(Tuple *tuple, RID *rid) -> bool override;
+
+  /** @return The output schema for the sequential scan */
```
Comment (lmwnshn) [inline src/include/execution/executors/filter_executor.h:49]:
```suggestion
  /** @return The output schema for the filter */
```

File: src/include/execution/executors/projection_executor.h [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+  /**
+   * Yield the next tuple from the projection.
+   * @param[out] tuple The next tuple produced by the projection
+   * @param[out] rid The next tuple RID produced by the projection
+   * @return `true` if a tuple was produced, `false` if there are no more tuples
+   */
+  auto Next(Tuple *tuple, RID *rid) -> bool override;
+
+  /** @return The output schema for the sequential scan */
```
Comment (lmwnshn) [inline src/include/execution/executors/projection_executor.h:49]:
```suggestion
  /** @return The output schema for the projection */
```

File: src/include/execution/executors/sort_executor.h [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+  /**
+   * Yield the next tuple from the sort.
+   * @param[out] tuple The next tuple produced by the sort
+   * @param[out] rid The next tuple RID produced by the sort
+   * @return `true` if a tuple was produced, `false` if there are no more tuples
+   */
+  auto Next(Tuple *tuple, RID *rid) -> bool override;
+
+  /** @return The output schema for the sequential scan */
```
Comment (lmwnshn) [inline src/include/execution/executors/sort_executor.h:48]:
```suggestion
  /** @return The output schema for the sort */
```

File: src/include/execution/executors/topn_executor.h [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+  /**
+   * Yield the next tuple from the topn.
+   * @param[out] tuple The next tuple produced by the topn
+   * @param[out] rid The next tuple RID produced by the topn
+   * @return `true` if a tuple was produced, `false` if there are no more tuples
+   */
+  auto Next(Tuple *tuple, RID *rid) -> bool override;
+
+  /** @return The output schema for the sequential scan */
```
Comment (lmwnshn) [inline src/include/execution/executors/topn_executor.h:48]:
```suggestion
  /** @return The output schema for the topn */
```

File: src/include/execution/executors/values_executor.h [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+   * @param exec_ctx The executor context
+   * @param plan The sequential scan plan to be executed
```
Comment (lmwnshn) [inline src/include/execution/executors/values_executor.h:33]:
```suggestion
   * @param plan The values plan to be executed
```

File: src/include/execution/executors/values_executor.h [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+  /**
+   * Yield the next tuple from the values.
+   * @param[out] tuple The next tuple produced by the values
+   * @param[out] rid The next tuple RID produced by the values
+   * @return `true` if a tuple was produced, `false` if there are no more tuples
+   */
+  auto Next(Tuple *tuple, RID *rid) -> bool override;
+
+  /** @return The output schema for the sequential scan */
```
Comment (lmwnshn) [inline src/include/execution/executors/values_executor.h:48]:
```suggestion
  /** @return The output schema for the values */
```

File: src/include/execution/plans/projection_plan.h [hunk 17]
```diff
@@ -0,0 +1,59 @@
+//===----------------------------------------------------------------------===//
+//
+//                         BusTub
+//
+// seq_scan_plan.h
```
Comment (lmwnshn) [inline src/include/execution/plans/projection_plan.h:5]:
```suggestion
// projection_plan.h
```

File: src/include/execution/plans/projection_plan.h [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+//
+// Identification: src/include/execution/plans/seq_scan_plan.h
```
Comment (lmwnshn) [inline src/include/execution/plans/projection_plan.h:7]:
```suggestion
// Identification: src/include/execution/plans/projection_plan.h
```

File: src/include/execution/plans/sort_plan.h [hunk 19]
```diff
@@ -0,0 +1,52 @@
+//===----------------------------------------------------------------------===//
+//
+//                         BusTub
+//
+// seq_scan_plan.h
```
Comment (lmwnshn) [inline src/include/execution/plans/sort_plan.h:5]:
```suggestion
// sort_plan.h
```

File: src/include/execution/plans/sort_plan.h [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+//
+// Identification: src/include/execution/plans/seq_scan_plan.h
```
Comment (lmwnshn) [inline src/include/execution/plans/sort_plan.h:7]:
```suggestion
// Identification: src/include/execution/plans/sort_plan.h
```

File: src/include/execution/plans/topn_plan.h [hunk 21]
```diff
@@ -0,0 +1,52 @@
+//===----------------------------------------------------------------------===//
+//
+//                         BusTub
+//
+// seq_scan_plan.h
```
Comment (lmwnshn) [inline src/include/execution/plans/topn_plan.h:5]:
```suggestion
// topn_plan.h
```

File: src/include/execution/plans/topn_plan.h [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+//
+// Identification: src/include/execution/plans/seq_scan_plan.h
```
Comment (lmwnshn) [inline src/include/execution/plans/topn_plan.h:7]:
```suggestion
// Identification: src/include/execution/plans/topn_plan.h
```

File: src/include/execution/plans/values_plan.h [hunk 23]
```diff
@@ -0,0 +1,47 @@
+//===----------------------------------------------------------------------===//
+//
+//                         BusTub
+//
+// seq_scan_plan.h
```
Comment (lmwnshn) [inline src/include/execution/plans/values_plan.h:5]:
```suggestion
// values_plan.h
```

File: src/include/execution/plans/values_plan.h [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+//
+// Identification: src/include/execution/plans/seq_scan_plan.h
```
Comment (lmwnshn) [inline src/include/execution/plans/values_plan.h:7]:
```suggestion
// Identification: src/include/execution/plans/values_plan.h
```

File: test/binder/binder_test.cpp [hunk 25]
```diff
@@ -141,6 +147,8 @@ TEST(BinderTest, DISABLED_BindInsert) {
   TryBind("INSERT INTO y SELECT * FROM y WHERE x < 500");
 }
 
+TEST(BinderTest, BindVarchar) { TryBind(R"(INSERT INTO c VALUES ('1', '2'))"); }
```
Comment (lmwnshn) [inline test/binder/binder_test.cpp:150]: Maybe try (1) a longer varchar and (2) an empty varchar too?

File: src/planner/plan_expression.cpp [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+    // Projections, Filters, and other executors evaluating expressions with one single child will
+    // use this branch.
+    const auto &child = children[0];
+    auto schema = child->OutputSchema();
+    uint32_t col_idx = schema->GetColIdx(col_name);
+    auto col_type = schema->GetColumn(col_idx).GetType();
+    return std::make_tuple(col_name, std::make_unique<ColumnValueExpression>(0, col_idx, col_type));
+  }
+  if (children.size() == 2) {
+    /**
```
Comment (lmwnshn) [inline src/planner/plan_expression.cpp:77]:
```suggestion
    /*
```

Comment (lmwnshn) [reply src/planner/plan_expression.cpp:77]: `/**` is for doxygen function blocks, general block comments can stick with `/*`

File: src/planner/planner.cpp [hunk 27]
```diff
@@ -137,14 +66,15 @@ auto Planner::PlanSelect(const SelectStatement &statement) -> std::unique_ptr<Ab
 
   if (!statement.where_->IsInvalid()) {
     auto schema = plan->OutputSchema();
-    auto expr = PlanExpression(*statement.where_, {&*plan});
+    auto [_, expr] = PlanExpression(*statement.where_, {&*plan});
```
Comment (lmwnshn) [inline src/planner/planner.cpp:69]:
```suggestion
    auto [_, expr] = PlanExpression(*statement.where_, {plan.get()});
```

if possible?

Comment (lmwnshn) [reply src/planner/planner.cpp:69]: same for other usages below

**Review by lmwnshn — APPROVED** [review]
LGTM, minor nits

File: src/include/execution/executors/filter_executor.h (hunk 2 above)
Comment (skyzh) [reply src/include/execution/executors/filter_executor.h:31]: My fault again 🤣

File: src/planner/planner.cpp (hunk 27 above)
Comment (skyzh) [reply src/planner/planner.cpp:69]: 😴 I'll change this when I wake up.

**Pull request merged**